plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryMetadata.java

```diff
@@ -31,4 +39,90 @@ public void testDatasetNotFoundMessage()
                 .matches(e -> e.getCode() == 404 && e.getMessage().contains("Not found: Dataset"));
     }
+
+    @Test
+    public void testProjectParentColumnsSingleParent()
+    {
+        BigQueryColumnHandle parentColumn = testingColumn("a", ImmutableList.of());
+        List<BigQueryColumnHandle> columns = ImmutableList.of(
+                parentColumn,
+                testingColumn("a", ImmutableList.of("b")),
+                testingColumn("a", ImmutableList.of("b", "c")),
+                testingColumn("a", ImmutableList.of("b", "c", "d")));
+
+        List<BigQueryColumnHandle> parentColumns = projectParentColumns(columns);
+        assertThat(parentColumns).size().isEqualTo(1);
+        assertThat(parentColumns.getFirst()).isEqualTo(parentColumn);
+    }
+
+    @Test
+    public void testProjectParentColumnsSingleParentDifferentOrder()
+    {
+        BigQueryColumnHandle parentColumn = testingColumn("a", ImmutableList.of());
+        List<BigQueryColumnHandle> columns = ImmutableList.of(
+                parentColumn,
+                testingColumn("a", ImmutableList.of("b")),
+                testingColumn("a", ImmutableList.of("b", "c")),
+                testingColumn("a", ImmutableList.of("b", "c", "d")),
+                testingColumn("a", ImmutableList.of("d", "c", "b")));
+
+        List<BigQueryColumnHandle> parentColumns = projectParentColumns(columns);
+        assertThat(parentColumns).size().isEqualTo(1);
+        assertThat(parentColumns.getFirst()).isEqualTo(parentColumn);
+    }
+
+    @Test
+    public void testProjectParentColumnsNoParentDifferentOrder()
+    {
+        List<BigQueryColumnHandle> columns = ImmutableList.of(
+                testingColumn("a", ImmutableList.of("b", "c", "d")),
+                testingColumn("a", ImmutableList.of("d", "c", "b")));
+
+        List<BigQueryColumnHandle> parentColumns = projectParentColumns(columns);
+        assertThat(parentColumns).size().isEqualTo(2);
+    }
+
+    @Test
+    public void testProjectParentColumnsSingleParentSuddenJump()
+    {
+        BigQueryColumnHandle parentColumn = testingColumn("a", ImmutableList.of());
+        List<BigQueryColumnHandle> columns = ImmutableList.of(
+                parentColumn,
+                testingColumn("a", ImmutableList.of("d", "c", "b")));
+
+        List<BigQueryColumnHandle> parentColumns = projectParentColumns(columns);
+        assertThat(parentColumns).size().isEqualTo(1);
+        assertThat(parentColumns.getFirst()).isEqualTo(parentColumn);
+    }
+
+    @Test
+    public void testProjectParentColumnsMultipleParent()
+    {
+        BigQueryColumnHandle parentColumn = testingColumn("a", ImmutableList.of());
+        BigQueryColumnHandle anotherParentColumn = testingColumn("a1", ImmutableList.of());
+        List<BigQueryColumnHandle> columns = ImmutableList.of(
+                parentColumn,
+                anotherParentColumn,
+                testingColumn("a", ImmutableList.of("b")),
```

Comment:
What if we could remove this column ? i.e if the dereferences are not increased in a incremental way

Reply:
And also lets have the order in a random way

Reply:
added